### PR TITLE
[COART-181]: Check Gracefully termination in all our applications

### PR DESCRIPTION
**Pull request by frnmjn**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/COART-181

More details on how to use in the app https://github.com/primait/peano/pull/2425

### CODE REVIEW

File: lib/amqp/gen/consumer.ex [hunk 1]
```diff
@@ -26,6 +27,13 @@ defmodule Amqpx.Gen.Consumer do
 
   @gen_server_opts [:name, :timeout, :debug, :spawn_opt, :hibernate_after]
 
+  defp signal_handler do
```
Comment (cpiemontese) [inline lib/amqp/gen/consumer.ex:30]: Super minor nit but I'd move this down near the other "internal" funs

File: lib/amqp/gen/consumer.ex [hunk 2]
```diff
@@ -242,4 +259,21 @@ defmodule Amqpx.Gen.Consumer do
...
+  defp cancel?(%{cancel?: true} = state, _), do: state
+
+  defp cancel?(state, consumer_tag) do
+    Logger.info("Cancelling consumer #{consumer_tag}")
+
+    if draining?() do
+      Basic.cancel(state.channel, consumer_tag)
+    end
+
+    %{state | cancel?: true}
```
Comment (cpiemontese) [inline lib/amqp/gen/consumer.ex:277]: If `cancel?` is to mark that it's cancelling then I would call this `cancelling?` to match the other vars, otherwise I misunderstood the logic

File: lib/amqp/gen/consumer.ex [hunk 3]
```diff
@@ -211,17 +221,24 @@ defmodule Amqpx.Gen.Consumer do
...
            handler_module: handler_module,
            handler_state: handler_state,
            backoff: backoff,
            requeue_on_reject: requeue_on_reject
          } = state
        ) do
-    {:ok, handler_state} = handler_module.handle_message(message, meta, handler_state)
-    Basic.ack(state.channel, tag)
-    %{state | handler_state: handler_state}
+    state = cancel?(state, consumer_tag)
```
Comment (cpiemontese) [inline lib/amqp/gen/consumer.ex:232]: If this is called with `cancel? = false` won't it trigger cancelling at the first `handle_message`?

File: lib/amqp/no_signal_handler.ex [hunk 4]
```diff
@@ -0,0 +1,13 @@
+defmodule Amqpx.NoSignalHandler do
+  @moduledoc """
+  Dummy signal handler module that does handle the graceful termination.
```
Comment (cpiemontese) [inline lib/amqp/no_signal_handler.ex:3]:
```suggestion
  Dummy signal handler module that handles the graceful termination.
```

**Review by cpiemontese — COMMENTED** [review]
A bit unclear on the logic

**Review by cpiemontese — DISMISSED** [review]
I think it makes more sense now, basically if the SignalHandler changes state, then processing may transition to draining and eventually stopping

**Review by ivanopagano — APPROVED** [review]
Looks fine

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ---
 
+## [6.1.0] - 2024-11-29
+
+### Added
+
+- ([#208](https://github.com/primait/amqpx/pull/208)) Introduce the graceful termination using the application signal handler to cancel the consumers on draining phase and close the channels on stopping phase.
```
Comment (cpiemontese) [inline CHANGELOG.md:16]: Not to be too nitpick, but technically the PR introduces the possibility of configuring a signal handler which can be used for graceful termination and so on, right?

If so, I would rephrase this a bit so expectations are clearer

**Review by cpiemontese — APPROVED** [review]
